### PR TITLE
[GraphTrainer][AutoDev] Introduce CPU offload pass for activation memory savings

### PR DESCRIPTION
**Pull request by SherlockNoMad**
## Summary

- Add a graph-level CPU offload pass (`cpu_offload.py`) that asynchronously transfers activation tensors to CPU during forward and reloads them during backward, reducing peak GPU memory usage
- Define custom `ao::offload`, `ao::reload`, and `ao::wait_tensor` ops for async GPU<->CPU transfers with event-based synchronization on a dedicated transfer stream
- Integrate the pass into `apply_default_graph_passes` (make_fx path) and `AVAILABLE_JOINT_PASSES` registry (AOT path)

### Key design decisions

- Uses `meta["custom"]["ac_region_id"]` (from `annotate_ac_regions`) as layer boundary markers, and `seq_nr` metadata for forward/backward node classification -- adapting the internal reference (D99042273) to work with make_fx traced graphs
- Skips view ops, small tensors (<4KB), collectives, and the last layer
- Is a no-op when no nodes are tagged (e.g., without `ac_region_id` annotations), so existing workflows are unaffected
- Handles symbolic FakeTensors (DSv3 MoE) gracefully

### Scheduling optimizations (deferred)

The make_fx traced joint graph interleaves forward and backward nodes (unlike manually built graphs where they are cleanly separated). The defer/prefetch scheduling optimizations from the reference implementation need adaptation for this interleaved structure. For now, offload/reload ops are placed adjacent to their producers/consumers, which still saves memory but without async D2H/H2D overlap with compute. This will be addressed in a follow-up.

### View ops

View op support is also deferred to a follow-up.

## GPU validation (8xH100, FSDP4+TP2)

**Llama3 debug model:**
- 40 activations offloaded (61 MB)
- Loss: 8.08 -> 5.58 over 5 steps
- Memory: 0.80 GiB (0.84%)

**DeepSeek-v3 debug model (EP4):**
- 112 activations offloaded (6.5 GB)
- Loss: 8.18 -> 4.53 over 5 steps
- Memory: 15.23 GiB (16.03%)

## Test plan

- [x] 7 new unit tests in `test_passes.py::TestCpuOffloadPass` covering:
  - Tagging correctly identifies offloadable nodes across multiple layers
  - Last layer is excluded from tagging
  - Nodes without backward consumers are not tagged
  - View ops are not tagged
  - Small tensors below threshold are not tagged
  - Offload/reload/wait ops are inserted in correct positions
  - Pass is a no-op when no nodes are tagged
- [x] All 19 tests in `test_passes.py` pass (12 existing + 7 new)
- [x] Bitwise deterministic test passes (Llama3 + DSv3)
- [x] Llama3 debug model training (8 GPU, FSDP4+TP2) - 5 steps, loss converges
- [x] DSv3 debug model training (8 GPU, FSDP4+TP2+EP4) - 5 steps, loss converges

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+# Custom ops for async activation offloading (ao:: namespace)
+#
+# A single dedicated transfer stream handles all D2H/H2D copies.
+# Completion events are keyed by output tensor data_ptr() and
+# cached for reuse across CUDA graph replays.
+#
+# ``keepalive`` in the offload wait_tensor creates a graph dependency
+# that extends the GPU tensor's lifetime past the async D2H copy,
+# replacing ``record_stream`` (which causes memory fragmentation).
+# ============================================================
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: you should use the custom op from upstream pytorch  ao::offload, ao::reload, ao::wait ops for activation offloading. See https://github.com/pytorch/pytorch/pull/177621.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+        f"({total_bytes / 1024 / 1024:.2f} MB)"
+    )
+
+
+# ============================================================
+# Graph passes
+# ============================================================
+
+
+def _tlparse_log_graph(gm: torch.fx.GraphModule, graph_name: str) -> None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: remove this. already have one in passes.py.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+        torch.ops._c10d_functional.reduce_scatter_tensor,
+        torch.ops._c10d_functional.all_to_all_single,
+        torch.ops._c10d_functional.wait_tensor,
+    }
+)
+
+_MIN_OFFLOAD_BYTES = 4096  # 4 KB minimum to justify offload overhead
+
+
+class OffloadPolicy(enum.Enum):
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: Does upstream pytorch have a equivalent enum? from https://github.com/pytorch/pytorch/pull/177621

If so, use that.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+
+        # Move reload to just before the target layer's backward start
+        target.prepend(reload_node)
+        moved += 1
+
+    if moved:
+        logger.info(f"CPU offload prefetch: moved {moved} reload ops")
+
+
+def _apply_cpu_offload_pass(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: make this a public function, ensure it follows the graph pass signature.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+
+    return forward_nodes, backward_nodes
+
+
+# ============================================================
+# Tagging pass
+# ============================================================
+
+
+def tag_offloadable_activations(gm: torch.fx.GraphModule) -> None:
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: find a better name for this function. 
This is offloading all offloadable activation.
Mention that this is a reference implementation. In practice, use should tag activations via annotation or graph passes.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+        return False
+    return True
+
+
+# ============================================================
+# Forward/backward node classification for make_fx traced graphs
+# ============================================================
+
+
+def _classify_forward_backward(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py:171]: TODO: there should a standard way to distinguish forward nodes from backward nodes. 
The backward tag should come from upstream pytorch autograd engine. 

cc @tugsbayasgalan

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+            continue
+        layer_id = node.meta.get("custom", {}).get(_AC_REGION_ID)
+        if layer_id is None:
+            continue
+        if layer_id not in bwd_start:
+            bwd_start[layer_id] = node
+    return bwd_start
+
+
+def _defer_offload_waits(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: not used, remove. 
Create a backlog task to introduce defer offload wait, mention it's dependecy on this PR lands.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 8]
```diff
@@ -0,0 +1,736 @@
...
+        target.append(wait_node)
+        moved += 1
+
+    if moved:
+        logger.info(
+            f"CPU offload defer: moved {moved} offload wait ops to end of next layer"
+        )
+
+
+def _prefetch_reloads(
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: not used, remove. 
create a backlog task to introduce prefetch, mention it's dependecy on this PR lands.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 9]
```diff
@@ -0,0 +1,736 @@
...
+        target.prepend(reload_node)
+        moved += 1
+
+    if moved:
+        logger.info(f"CPU offload prefetch: moved {moved} reload ops")
+
+
+def _apply_cpu_offload_pass(
+    gm: torch.fx.GraphModule,
+    device: torch.device,
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: you don't really need this argument. 
you should be able to infer the reload device target from the original tensor's device before offload.

File: torchtitan/experiments/graph_trainer/cpu_offload.py [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+        f"({total_bytes / 1024 / 1024:.2f} MB)"
+    )
+    return gm
+
+
+def cpu_offload_pass(
+    gm: torch.fx.GraphModule,
+    example_inputs: tuple | None = None,
+    *,
+    device: torch.device | None = None,
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/cpu_offload.py]: same here, you don't need device as arg.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Replaced all custom op definitions (offload, reload, wait_tensor, stream management, event registry) with an import of the upstream `torch._functorch._activation_offloading.offload_ops` module. The ops are now registered as a side-effect of importing that module.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Removed the duplicate `_tlparse_log_graph` helper. Now using `tlparse_log_graph_pass` from `passes.py` instead (lazy import to avoid circular dependency).

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 3 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Yes, upstream pytorch has `CheckpointPolicy.MUST_CPU_OFFLOAD` in `torch.utils.checkpoint`. Replaced the custom `OffloadPolicy` enum with `CheckpointPolicy.MUST_CPU_OFFLOAD` throughout.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 4 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Renamed `_apply_cpu_offload_pass` to `apply_cpu_offload_pass` (public) and updated its signature to follow the graph pass convention: `(gm: GraphModule, example_inputs: tuple | None = None) -> GraphModule`.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 5 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Renamed to `tag_all_offloadable_activations` to better reflect that it offloads all offloadable activations. Added a docstring noting this is a reference implementation and that in practice users should tag activations via annotations or graph passes for more fine-grained control.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 6 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py:171]: AutoDev: Added a TODO comment above `_classify_forward_backward` noting that there should be a standard way to distinguish forward nodes from backward nodes, and that the backward tag should come from upstream pytorch autograd engine. (cc @tugsbayasgalan)

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 7 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Removed `_defer_offload_waits`. Created a Backlog draft issue on the project board: "Introduce defer offload wait for CPU activation offloading" with dependency on this PR landing first.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 8 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Removed `_prefetch_reloads`. Created a Backlog draft issue on the project board: "Introduce activation prefetch for CPU offload" with dependency on this PR landing first.

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 9 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Removed the `device` argument from `apply_cpu_offload_pass`. The reload device is now inferred from `node.meta["val"].device` (the original tensor's device before offloading).

File: torchtitan/experiments/graph_trainer/cpu_offload.py (hunk 10 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/cpu_offload.py]: AutoDev: Removed the `device` argument from `cpu_offload_pass` as well. The device inference now happens per-tensor inside `apply_cpu_offload_pass` from `node.meta["val"].device`.
